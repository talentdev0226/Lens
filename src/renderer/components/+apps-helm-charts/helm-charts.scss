/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

.HelmCharts {
  .SearchInput {
    width: 70%;
    margin: auto;

    > label {
      padding: $padding $padding * 2;
    }
  }

  .TableCell {
    text-overflow: ellipsis;

    &.name {
      flex-grow: 1.3;
    }

    &.icon {
      $iconSize: $unit * 3.5;

      padding: 0;
      display: flex;
      flex-grow: 0;
      flex-basis: $iconSize;

      figure {
        width: $iconSize;
        height: $iconSize;
        border-radius: 50%;
        background: var(--helmImgBackground) url(helm-placeholder.svg) center center no-repeat;
        background-size: 72%; // bg size looks same as image on top of it
        margin: auto;

        img {
          object-fit: contain;
          width: inherit;
          height: inherit;
          visibility: hidden;
          border-radius: inherit;
          background-color: var(--helmImgBackground);
          padding: $padding * 0.5;

          &.visible {
            visibility: visible;
          }
        }
      }
    }

    &.description {
      flex-grow: 2.5;
    }

    &.repository {
      &.stable {
        color: var(--helmStableRepo);
      }

      &.incubator {
        color: var(--helmIncubatorRepo);
      }
    }
  }
}
