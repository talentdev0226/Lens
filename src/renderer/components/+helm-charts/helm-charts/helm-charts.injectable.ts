/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import { asyncComputed } from "@ogre-tools/injectable-react";
import callForHelmChartsInjectable from "./call-for-helm-charts.injectable";

const helmChartsInjectable = getInjectable({
  id: "helm-charts",

  instantiate: (di) => {
    const callForHelmCharts = di.inject(callForHelmChartsInjectable);

    return asyncComputed(async () => await callForHelmCharts(), []);
  },
});

export default helmChartsInjectable;
