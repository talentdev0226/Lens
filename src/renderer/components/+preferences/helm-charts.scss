.HelmCharts {
  .repos {
    margin-top: var(--margin);

    .Badge {
      display: flex;
      margin-bottom: 1px!important;
      padding: 6px 8px;
    }
  }
}