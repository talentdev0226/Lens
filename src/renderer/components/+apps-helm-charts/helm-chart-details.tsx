import "./helm-chart-details.scss";

import React, { Component } from "react";
import { getChartDetails, HelmChart } from "../../api/endpoints/helm-charts.api";
import { observable, autorun } from "mobx";
import { observer } from "mobx-react";
import { Drawer, DrawerItem } from "../drawer";
import { autobind, stopPropagation } from "../../utils";
import { MarkdownViewer } from "../markdown-viewer";
import { Spinner } from "../spinner";
import { Button } from "../button";
import { Select, SelectOption } from "../select";
import { createInstallChartTab } from "../dock/install-chart.store";
import { Badge } from "../badge";

interface Props {
  chart: HelmChart;
  hideDetails(): void;
}

@observer
export class HelmChartDetails extends Component<Props> {
  @observable chartVersions: HelmChart[];
  @observable selectedChart: HelmChart;
  @observable readme: string = null;
  @observable error: string = null;

  private abortController?: AbortController;

  componentWillUnmount() {
    this.abortController?.abort();
  }

  chartUpdater = autorun(() => {
    this.selectedChart = null;
    const { chart: { name, repo, version } } = this.props;

    getChartDetails(repo, name, { version })
      .then(result => {
        this.readme = result.readme;
        this.chartVersions = result.versions;
        this.selectedChart = result.versions[0];
      })
      .catch(error => {
        this.error = error;
      });
  });

  @autobind()
  async onVersionChange({ value: version }: SelectOption<string>) {
    this.selectedChart = this.chartVersions.find(chart => chart.version === version);
    this.readme = null;

    try {
      this.abortController?.abort();
      this.abortController = new AbortController();
      const { chart: { name, repo } } = this.props;
      const { readme } = await getChartDetails(repo, name, { version, reqInit: { signal: this.abortController.signal }});

      this.readme = readme;
    } catch (error) {
      this.error = error;
    }
  }

  @autobind()
  install() {
    createInstallChartTab(this.selectedChart);
    this.props.hideDetails();
  }

  renderIntroduction() {
    const { selectedChart, chartVersions, onVersionChange } = this;
    const placeholder = require("./helm-placeholder.svg");

    return (
      <div className="introduction flex align-flex-start">
        <img
          className="intro-logo"
          src={selectedChart.getIcon() || placeholder}
          onError={(event) => event.currentTarget.src = placeholder}
        />
        <div className="intro-contents box grow">
          <div className="description flex align-center justify-space-between">
            {selectedChart.getDescription()}
            <Button primary label="Install" onClick={this.install} />
          </div>
          <DrawerItem name="Version" className="version" onClick={stopPropagation}>
            <Select
              themeName="outlined"
              menuPortalTarget={null}
              options={chartVersions.map(chart => chart.version)}
              value={selectedChart.getVersion()}
              onChange={onVersionChange}
            />
          </DrawerItem>
          <DrawerItem name="Home">
            <a href={selectedChart.getHome()} target="_blank" rel="noreferrer">{selectedChart.getHome()}</a>
          </DrawerItem>
          <DrawerItem name="Maintainers" className="maintainers">
            {selectedChart.getMaintainers().map(({ name, email, url }) =>
              <a key={name} href={url || `mailto:${email}`} target="_blank" rel="noreferrer">{name}</a>
            )}
          </DrawerItem>
          {selectedChart.getKeywords().length > 0 && (
            <DrawerItem name="Keywords" labelsOnly>
              {selectedChart.getKeywords().map(key => <Badge key={key} label={key} />)}
            </DrawerItem>
          )}
        </div>
      </div>
    );
  }

  renderReadme() {
    if (this.readme === null) {
      return <Spinner center />;
    }

    return (
      <div className="chart-description">
        <MarkdownViewer markdown={this.readme} />
      </div>
    );
  }

  renderContent() {
    if (!this.selectedChart) {
      return <Spinner center />;
    }

    if (this.error) {
      return (
        <div className="box grow">
          <p className="error">{this.error}</p>
        </div>
      );
    }

    return (
      <div className="box grow">
        {this.renderIntroduction()}
        {this.renderReadme()}
      </div>
    );
  }

  render() {
    const { chart, hideDetails } = this.props;
    const title = chart ? <>Chart: {chart.getFullName()}</> : "";

    return (
      <Drawer
        className="HelmChartDetails"
        usePortal={true}
        open={!!chart}
        title={title}
        onClose={hideDetails}
      >
        {this.renderContent()}
      </Drawer>
    );
  }
}
