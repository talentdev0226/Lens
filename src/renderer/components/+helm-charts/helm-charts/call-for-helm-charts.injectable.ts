/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import type { HelmChart } from "../../../../common/k8s-api/endpoints/helm-charts.api";
import { listCharts } from "../../../../common/k8s-api/endpoints/helm-charts.api";

export type CallForHelmCharts = () => Promise<HelmChart[]>;

const callForHelmChartsInjectable = getInjectable({
  id: "call-for-helm-charts",
  instantiate: (): CallForHelmCharts => async () => await listCharts(),
  causesSideEffects: true,
});

export default callForHelmChartsInjectable;
