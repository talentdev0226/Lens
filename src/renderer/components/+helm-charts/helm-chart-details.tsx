/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./helm-chart-details.scss";

import React, { Component } from "react";
import type { HelmChart } from "../../../common/k8s-api/endpoints/helm-charts.api";
import { computed, observable, reaction, runInAction } from "mobx";
import { disposeOnUnmount, observer } from "mobx-react";
import { Drawer, DrawerItem } from "../drawer";
import { autoBind, stopPropagation } from "../../utils";
import { MarkdownViewer } from "../markdown-viewer";
import { Spinner } from "../spinner";
import { Button } from "../button";
import { Select } from "../select";
import { Badge } from "../badge";
import { Tooltip, withStyles } from "@material-ui/core";
import { withInjectables } from "@ogre-tools/injectable-react";
import createInstallChartTabInjectable from "../dock/install-chart/create-install-chart-tab.injectable";
import type { ShowCheckedErrorNotification } from "../notifications/show-checked-error.injectable";
import type { SingleValue } from "react-select";
import AbortController from "abort-controller";
import showCheckedErrorNotificationInjectable from "../notifications/show-checked-error.injectable";
import type { GetChartDetails } from "./get-char-details.injectable";
import getChartDetailsInjectable from "./get-char-details.injectable";
import { HelmChartIcon } from "./icon";

export interface HelmChartDetailsProps {
  chart: HelmChart;
  hideDetails(): void;
}

const LargeTooltip = withStyles({
  tooltip: {
    fontSize: "var(--font-size-small)",
  },
})(Tooltip);

interface Dependencies {
  createInstallChartTab: (helmChart: HelmChart) => void;
  showCheckedErrorNotification: ShowCheckedErrorNotification;
  getChartDetails: GetChartDetails;
}

@observer
class NonInjectedHelmChartDetails extends Component<HelmChartDetailsProps & Dependencies> {
  readonly chartVersions = observable.array<HelmChart>();
  readonly selectedChart = observable.box<HelmChart | undefined>();
  readonly readme = observable.box<string | undefined>(undefined);
  readonly chartVerionOptions = computed(() => (
    this.chartVersions.map(chart => ({
      value: chart,
      label: chart.version,
    }))
  ));

  private abortController = new AbortController();

  constructor(props: HelmChartDetailsProps & Dependencies) {
    super(props);
    autoBind(this);
  }

  componentWillUnmount() {
    this.abortController.abort();
  }

  componentDidMount() {
    disposeOnUnmount(this, [
      reaction(() => this.props.chart, async ({ name, repo, version }) => {
        runInAction(() => {
          this.selectedChart.set(undefined);
          this.chartVersions.clear();
          this.readme.set("");
        });

        try {
          const { readme, versions } = await this.props.getChartDetails(repo, name, { version });

          runInAction(() => {
            this.readme.set(readme);
            this.chartVersions.replace(versions);
            this.selectedChart.set(versions[0]);
          });
        } catch (error) {
          this.props.showCheckedErrorNotification(error, "Unknown error occured while getting chart details");
        }
      }, {
        fireImmediately: true,
      }),
    ]);
  }

  async onVersionChange(option: SingleValue<{ value: HelmChart }>) {
    const chart = option?.value ?? this.chartVersions[0];

    runInAction(() => {
      this.selectedChart.set(chart ?? undefined);
      this.readme.set(undefined);
    });

    try {
      this.abortController.abort();
      this.abortController = new AbortController();
      const { chart: { name, repo }} = this.props;
      const { readme } = await this.props.getChartDetails(repo, name, { version: chart.version, reqInit: { signal: this.abortController.signal }});

      this.readme.set(readme);
    } catch (error) {
      this.props.showCheckedErrorNotification(error, "Unknown error occured while getting chart details");
    }
  }

  install(selectedChart: HelmChart) {
    this.props.createInstallChartTab(selectedChart);
    this.props.hideDetails();
  }

  renderIntroduction(selectedChart: HelmChart) {
    return (
      <div className="introduction flex align-flex-start">
        <HelmChartIcon
          chart={selectedChart}
          className="intro-logo"
        />
        <div className="intro-contents box grow">
          <div className="description flex align-center justify-space-between" data-testid="selected-chart-description">
            {selectedChart.getDescription()}
            <Button
              primary
              label="Install"
              onClick={() => this.install(selectedChart)}
            />
          </div>
          <DrawerItem
            name="Version"
            className="version"
            onClick={stopPropagation}
          >
            <Select
              id="chart-version-input"
              themeName="outlined"
              menuPortalTarget={null}
              options={this.chartVerionOptions.get()}
              formatOptionLabel={({ value: chart }) => (
                chart.deprecated
                  ? (
                    <LargeTooltip title="Deprecated" placement="left">
                      <span className="deprecated">{chart.version}</span>
                    </LargeTooltip>
                  )
                  : chart.version
              )}
              isOptionDisabled={({ value: chart }) => chart.deprecated}
              value={selectedChart}
              onChange={this.onVersionChange}
            />
          </DrawerItem>
          <DrawerItem name="Home">
            <a
              href={selectedChart.getHome()}
              target="_blank"
              rel="noreferrer"
            >
              {selectedChart.getHome()}
            </a>
          </DrawerItem>
          <DrawerItem name="Maintainers" className="maintainers">
            {selectedChart.getMaintainers().map(({ name, email, url }) => (
              <a
                key={name}
                href={url || `mailto:${email}`}
                target="_blank"
                rel="noreferrer"
              >
                {name}
              </a>
            ))}
          </DrawerItem>
          {selectedChart.getKeywords().length > 0 && (
            <DrawerItem name="Keywords" labelsOnly>
              {selectedChart.getKeywords().map(key => <Badge key={key} label={key} />)}
            </DrawerItem>
          )}
        </div>
      </div>
    );
  }

  renderReadme() {
    const readme = this.readme.get();

    if (readme === undefined) {
      return <Spinner center />;
    }

    return (
      <div className="chart-description" data-testid="helmchart-readme">
        <MarkdownViewer markdown={readme} />
      </div>
    );
  }

  renderContent() {
    const selectedChart = this.selectedChart.get();

    if (!selectedChart) {
      return <Spinner center />;
    }

    return (
      <div className="box grow">
        {this.renderIntroduction(selectedChart)}
        {this.renderReadme()}
      </div>
    );
  }

  render() {
    const { chart, hideDetails } = this.props;

    return (
      <Drawer
        className="HelmChartDetails"
        usePortal={true}
        open={!!chart}
        title={chart ? `Chart: ${chart.getFullName()}` : ""}
        onClose={hideDetails}
      >
        {this.renderContent()}
      </Drawer>
    );
  }
}

export const HelmChartDetails = withInjectables<Dependencies, HelmChartDetailsProps>(NonInjectedHelmChartDetails, {
  getProps: (di, props) => ({
    ...props,
    createInstallChartTab: di.inject(createInstallChartTabInjectable),
    showCheckedErrorNotification: di.inject(showCheckedErrorNotificationInjectable),
    getChartDetails: di.inject(getChartDetailsInjectable),
  }),
});
