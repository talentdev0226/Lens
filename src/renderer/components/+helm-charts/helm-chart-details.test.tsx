/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import type { DiContainer } from "@ogre-tools/injectable";
import type { RenderResult } from "@testing-library/react";
import React from "react";
import directoryForLensLocalStorageInjectable from "../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import { HelmChart } from "../../../common/k8s-api/endpoints/helm-charts.api";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import { noop } from "../../utils";
import type { CreateInstallChartTab } from "../dock/install-chart/create-install-chart-tab.injectable";
import createInstallChartTabInjectable from "../dock/install-chart/create-install-chart-tab.injectable";
import { Notifications } from "../notifications";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import type { GetChartDetails } from "./get-char-details.injectable";
import getChartDetailsInjectable from "./get-char-details.injectable";
import { HelmChartDetails } from "./helm-chart-details";

describe("<HelmChartDetails />", () => {
  let di: DiContainer;
  let getChartDetails: AsyncFnMock<GetChartDetails>;
  let chart: HelmChart;
  let render: DiRender;
  let result: RenderResult;
  let createInstallChartTab: jest.MockedFunction<CreateInstallChartTab>;

  beforeEach(() => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });
    getChartDetails = asyncFn<GetChartDetails>();
    createInstallChartTab = jest.fn();
    chart = HelmChart.create({
      apiVersion: "some-api-version",
      created: "a long time ago",
      name: "a name",
      repo: "a galaxy far far away",
      version: "1",
    });

    di.override(directoryForLensLocalStorageInjectable, () => "some-directory-for-lens-local-storage");
    di.override(getChartDetailsInjectable, () => getChartDetails);
    di.override(createInstallChartTabInjectable, () => createInstallChartTab);
    render = renderFor(di);
    result = render((
      <>
        <HelmChartDetails chart={chart} hideDetails={noop} />
        <Notifications />
      </>
    ));
  });

  describe("before getChartDetails resolves", () => {
    it("renders", () => {
      expect(result.baseElement).toMatchSnapshot();
    });

    describe("when getChartDetails resolves with one version", () => {
      beforeEach(async () => {
        await getChartDetails.resolve({
          readme: "I am a readme",
          versions: [chart],
        });
      });

      it("renders", () => {
        expect(result.baseElement).toMatchSnapshot();
      });

      it("shows the readme", () => {
        expect(result.queryByTestId("helmchart-readme")).not.toBeNull();
      });

      it("shows the selected chart", () => {
        expect(result.queryByTestId("selected-chart-description")).not.toBeNull();
      });
    });

    describe("with getChartDetails rejects", () => {
      beforeEach(async () => {
        await getChartDetails.reject(new Error("some error"));
      });

      it("renders", () => {
        expect(result.baseElement).toMatchSnapshot();
      });
    });
  });
});
