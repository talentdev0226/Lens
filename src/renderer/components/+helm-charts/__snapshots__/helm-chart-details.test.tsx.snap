// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelmChartDetails /> before getChartDetails resolves renders 1`] = `
<body>
  <div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
  <div
    class="Animate slide-right Drawer HelmChartDetails right enter"
    style="--size: 725px; --enter-duration: 100ms; --leave-duration: 100ms;"
  >
    <div
      class="drawer-wrapper flex column"
    >
      <div
        class="drawer-title flex align-center"
      >
        <div
          class="drawer-title-text flex gaps align-center"
        >
          Chart: a galaxy far far away/a name
          <i
            class="Icon material interactive focusable"
            id="tooltip_target_1"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="content_copy"
            >
              content_copy
            </span>
            <div />
          </i>
        </div>
        <i
          class="Icon material interactive focusable"
          id="tooltip_target_2"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="close"
          >
            close
          </span>
          <div />
        </i>
      </div>
      <div
        class="drawer-content flex column box grow"
      >
        <div
          class="Spinner singleColor center"
        />
      </div>
    </div>
    <div
      class="ResizingAnchor horizontal leading"
    />
  </div>
</body>
`;

exports[`<HelmChartDetails /> before getChartDetails resolves when getChartDetails resolves with one version renders 1`] = `
<body>
  <div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
  <div
    class="Animate slide-right Drawer HelmChartDetails right enter"
    style="--size: 725px; --enter-duration: 100ms; --leave-duration: 100ms;"
  >
    <div
      class="drawer-wrapper flex column"
    >
      <div
        class="drawer-title flex align-center"
      >
        <div
          class="drawer-title-text flex gaps align-center"
        >
          Chart: a galaxy far far away/a name
          <i
            class="Icon material interactive focusable"
            id="tooltip_target_3"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="content_copy"
            >
              content_copy
            </span>
            <div />
          </i>
        </div>
        <i
          class="Icon material interactive focusable"
          id="tooltip_target_4"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="close"
          >
            close
          </span>
          <div />
        </i>
      </div>
      <div
        class="drawer-content flex column box grow"
      >
        <div
          class="box grow"
        >
          <div
            class="introduction flex align-flex-start"
          >
            <div
              class="intro-logo"
            >
              <svg
                viewBox="0 0 722.8 702"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="currentColor"
                >
                  <path
                    d="m318 299.5c2.1 1.6 4.8 2.5 7.6 2.5 6.9 0 12.6-5.5 12.9-12.3l.3-.2 4.3-76.7c-5.2.6-10.4 1.5-15.6 2.7-28.5 6.5-53.2 20.5-72.6 39.5l62.9 44.6z"
                  />
                  <path
                    d="m309.5 411.9c-1.4-5.9-6.6-9.9-12.4-10-.8 0-1.7.1-2.5.2l-.1-.2-75.5 12.8c11.7 32.2 33.4 58.5 60.8 76.1l29.2-70.7-.2-.3c1.1-2.4 1.4-5.2.7-7.9z"
                  />
                  <path
                    d="m284.4 357.5c2.5-.7 4.9-2.2 6.7-4.4 4.3-5.4 3.6-13.2-1.6-17.8l.1-.3-57.4-51.4c-17 27.8-25.1 61.1-21.4 95.3l73.6-21.2z"
                  />
                  <path
                    d="m340.2 380 21.2 10.2 21.1-10.1 5.3-22.9-14.6-18.2h-23.6l-14.6 18.2z"
                  />
                  <path
                    d="m384.2 289.4c.1 2.6 1 5.2 2.8 7.5 4.3 5.4 12.1 6.4 17.7 2.4l.2.1 62.5-44.3c-23.6-23.1-54.4-38.2-87.6-42.2z"
                  />
                  <path
                    d="m490.3 283.7-57.1 51.1v.2c-2 1.7-3.5 4.1-4.1 6.8-1.5 6.8 2.5 13.5 9.2 15.3l.1.3 74 21.3c1.6-16 .6-32.5-3.2-49-3.9-16.8-10.4-32.2-18.9-46z"
                  />
                  <path
                    d="m372.8 439.6c-1.2-2.3-3.2-4.3-5.8-5.5-2-.9-4-1.4-6-1.3-4.5.2-8.7 2.6-10.9 6.8h-.1l-37.1 67.1c25.7 8.8 54.1 10.7 82.5 4.2 5.1-1.2 10-2.5 14.9-4.2l-37.3-67.1z"
                  />
                  <path
                    d="m711.7 425-60.4-262.2c-3.2-13.7-12.5-25.3-25.3-31.4l-244.4-116.8c-7.1-3.4-14.8-4.9-22.7-4.5-6.2.3-12.3 1.9-17.9 4.5l-244.3 116.7c-12.8 6.1-22.1 17.7-25.3 31.4l-60.2 262.3c-2.8 12.2-.5 25 6.3 35.5.8 1.3 1.7 2.5 2.7 3.7l169.1 210.3c8.9 11 22.3 17.4 36.5 17.4l271.2-.1c14.2 0 27.7-6.4 36.5-17.4l169.1-210.3c8.9-10.9 12.2-25.4 9.1-39.1zm-93-3.2c-1.8 7.8-10.2 12.6-18.9 10.7-.1 0-.2 0-.2 0-.1 0-.2-.1-.3-.1-1.2-.3-2.7-.5-3.8-.8-5-1.3-8.6-3.3-13.1-5.1-9.7-3.5-17.7-6.4-25.5-7.5-4-.3-6 1.6-8.2 3-1.1-.2-4.4-.8-6.2-1.1-14 44-43.9 82.2-84.3 106.1.7 1.7 1.9 5.3 2.4 5.9-.9 2.5-2.3 4.8-1.1 8.6 2.8 7.4 7.4 14.6 13 23.2 2.7 4 5.4 7.1 7.8 11.7.6 1.1 1.3 2.8 1.9 3.9 3.8 8 1 17.3-6.2 20.8-7.3 3.5-16.3-.2-20.2-8.3-.6-1.1-1.3-2.7-1.8-3.8-2.1-4.7-2.8-8.8-4.2-13.4-3.3-9.7-6-17.8-10-24.6-2.2-3.3-5-3.7-7.5-4.5-.5-.8-2.2-4-3.1-5.6-8.1 3.1-16.4 5.6-25.1 7.6-37.9 8.6-75.9 5.1-109.9-7.9l-3.3 6c-2.5.7-4.8 1.3-6.3 3.1-5.3 6.4-7.5 16.6-11.3 26.3-1.5 4.6-2.1 8.7-4.2 13.4-.5 1.1-1.3 2.6-1.8 3.7-3.9 8.1-12.9 11.7-20.2 8.2-7.2-3.5-10-12.7-6.2-20.8.6-1.2 1.3-2.8 1.9-3.9 2.4-4.6 5.2-7.7 7.8-11.7 5.5-8.7 10.4-16.4 13.2-23.8.7-2.4-.3-5.8-1.3-8.3l2.7-6.4c-38.9-23.1-69.7-59.8-84.3-105.3l-6.4 1.1c-1.7-1-5.1-3.2-8.4-3-7.8 1.1-15.8 4-25.5 7.5-4.5 1.7-8.1 3.7-13.1 5-1.1.3-2.6.6-3.8.8-.1 0-.2.1-.3.1s-.2 0-.2 0c-8.7 1.9-17.1-2.9-18.9-10.7s3.8-15.7 12.4-17.8c.1 0 .2 0 .2-.1h.1c1.2-.3 2.8-.7 3.9-.9 5.1-1 9.2-.7 14-1.1 10.2-1.1 18.7-1.9 26.2-4.3 2.4-1 4.7-4.3 6.3-6.3l6.1-1.8c-6.9-47.5 4.8-94.2 29.8-131.9l-4.7-4.2c-.3-1.8-.7-6-2.9-8.4-5.8-5.4-13-9.9-21.8-15.3-4.2-2.4-8-4-12.1-7.1-.9-.7-2.1-1.7-3-2.4-.1-.1-.1-.1-.2-.2-7-5.6-8.6-15.2-3.6-21.6 2.8-3.6 7.2-5.3 11.7-5.2 3.5.1 7.1 1.4 10.2 3.8 1 .8 2.4 1.8 3.2 2.6 3.9 3.4 6.3 6.7 9.6 10.2 7.2 7.3 13.2 13.4 19.7 17.8 3.4 2 6.1 1.2 8.7.8.8.6 3.7 2.6 5.3 3.8 24.9-26.4 57.6-46 95.6-54.6 8.8-2 17.7-3.3 26.4-4.1l.3-6.2c1.9-1.9 4.1-4.6 4.8-7.6.6-7.9-.4-16.3-1.6-26.5-.7-4.8-1.8-8.7-2-13.9 0-1.1 0-2.5 0-3.8 0-.1 0-.3 0-.4 0-9 6.5-16.2 14.6-16.2s14.6 7.3 14.6 16.2c0 1.3.1 3 0 4.2-.2 5.2-1.3 9.1-2 13.9-1.2 10.2-2.3 18.7-1.7 26.5.6 3.9 2.9 5.5 4.8 7.3 0 1.1.2 4.6.3 6.5 46.5 4.1 89.7 25.4 121.4 58.7l5.6-4c1.9.1 6 .7 8.9-1 6.5-4.4 12.5-10.5 19.7-17.8 3.3-3.5 5.7-6.8 9.7-10.2.9-.8 2.3-1.8 3.2-2.6 7-5.6 16.8-5 21.8 1.3s3.4 16-3.6 21.6c-1 .8-2.3 1.9-3.2 2.6-4.2 3.1-8 4.7-12.2 7.1-8.7 5.4-16 9.9-21.8 15.3-2.7 2.9-2.5 5.7-2.8 8.3-.8.7-3.7 3.3-5.2 4.7 12.6 18.8 22.1 40.1 27.4 63.3 5.3 23.1 6.1 46.1 3.1 68.3l5.9 1.7c1.1 1.5 3.2 5.2 6.3 6.3 7.5 2.4 16 3.2 26.2 4.3 4.8.4 8.9.2 14 1.1 1.2.2 3 .7 4.2 1 8.9 2.4 14.4 10.4 12.6 18.2z"
                  />
                  <path
                    d="m428 401.7c-1-.2-2-.3-3-.2-1.7.1-3.3.5-4.9 1.3-6.2 3-9 10.4-6.2 16.7l-.1.1 29.6 71.4c28.5-18.2 49.8-45.3 61-76.6l-76.2-12.9z"
                  />
                </g>
              </svg>
            </div>
            <div
              class="intro-contents box grow"
            >
              <div
                class="description flex align-center justify-space-between"
                data-testid="selected-chart-description"
              >
                
                <button
                  class="Button primary"
                  type="button"
                >
                  Install
                </button>
              </div>
              <div
                class="DrawerItem version"
              >
                <span
                  class="name"
                >
                  Version
                </span>
                <span
                  class="value"
                >
                  <div
                    class="Select theme-outlined css-b62m3t-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-chart-version-input-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="Select__control css-1s2u09g-control"
                    >
                      <div
                        class="Select__value-container Select__value-container--has-value css-319lph-ValueContainer"
                      >
                        <div
                          class="Select__single-value css-qc6sy-singleValue"
                        >
                          1
                        </div>
                        <div
                          class="Select__input-container css-6j8wv5-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-haspopup="true"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class="Select__input"
                            id="chart-version-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="Select__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="Select__indicator-separator css-1okebmr-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="Select__indicator Select__dropdown-indicator css-tlfecz-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </div>
              <div
                class="DrawerItem"
              >
                <span
                  class="name"
                >
                  Home
                </span>
                <span
                  class="value"
                >
                  <a
                    rel="noreferrer"
                    target="_blank"
                  />
                </span>
              </div>
              <div
                class="DrawerItem maintainers"
              >
                <span
                  class="name"
                >
                  Maintainers
                </span>
                <span
                  class="value"
                />
              </div>
            </div>
          </div>
          <div
            class="chart-description"
            data-testid="helmchart-readme"
          >
            <div
              class="MarkDownViewer"
            >
              <p>
                I am a readme
              </p>
              

            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ResizingAnchor horizontal leading"
    />
  </div>
</body>
`;

exports[`<HelmChartDetails /> before getChartDetails resolves with getChartDetails rejects renders 1`] = `
<body>
  <div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex error enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Error: some error
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_20"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Animate slide-right Drawer HelmChartDetails right enter"
    style="--size: 725px; --enter-duration: 100ms; --leave-duration: 100ms;"
  >
    <div
      class="drawer-wrapper flex column"
    >
      <div
        class="drawer-title flex align-center"
      >
        <div
          class="drawer-title-text flex gaps align-center"
        >
          Chart: a galaxy far far away/a name
          <i
            class="Icon material interactive focusable"
            id="tooltip_target_18"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="content_copy"
            >
              content_copy
            </span>
            <div />
          </i>
        </div>
        <i
          class="Icon material interactive focusable"
          id="tooltip_target_19"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="close"
          >
            close
          </span>
          <div />
        </i>
      </div>
      <div
        class="drawer-content flex column box grow"
      >
        <div
          class="Spinner singleColor center"
        />
      </div>
    </div>
    <div
      class="ResizingAnchor horizontal leading"
    />
  </div>
</body>
`;
